### ISSUE TITLE
Proposal: Update Runge-Kutta step size algorithms for predictable performance

### ISSUE THREAD

**Issue by solter** [scipy/scipy#9822]
### Proposal
Update the adaptive step-size algorithm to capture the stability and conditioning improvements demonstrated in [1].

Feedback would be much appreciated on the technical details of the approach below, as well as on the overall worth-while-ness of this proposal.

### Context
In the past couple of decades the adaptive step-size algorithm has been studied and the stability and conditioning can be significantly improved without affecting speed. Here is an example where on the left a step-size algorithm similar to scipy's was used, and on the right was a new step-size algorithm based on control theory:
![workprecisiondiagrams](https://user-images.githubusercontent.com/5632706/52876852-fa05af80-311d-11e9-8198-585e36bd4664.jpg)
This was figure 1 from [1]. The right panel demonstrates that using more sophisticated step size algorithms from control theory result in much more predictable behavior in that increasing tolerance promises to *both* decrease accuracy *and* decrease runtime. This is in contrast to the left panel where increasing the tolerance will decrease accuracy but may not decrease runtime.

#### 1. Testing
The first step would be to add testing that accurately analyzes the current algorithms. This would be implemented following the analysis in [2]. This would involve 
- [ ] updating the method signatures to support the use of tolerance definitions described in [2]
- [ ] find/port a test suite of IVPs (maybe [Test Set for IVP Solvers](https://archimede.dm.uniba.it/~testset/testsetivpsolvers/))
- [ ] ❓  create test methods calculate summary stats for different parameters using RK45
  -  parameters
      - `lt` = log(tolerance)
      - `lf` = log(number of function calls)
      - `la` = log(achieved accuracy)
      -  runtime
  - summary stats
    - ❓ linear bias of `lf` vs `lt` (slope and intercept)
    - ❓ remove the linear bias from `lf` vs `lt`, then calculate the [unbiased sample variance](https://en.wikipedia.org/wiki/Variance#Sample_variance)
    - ❓ linear bias of `la` vs `lt` (slope and intercept)
    - ❓ remove the linear bias from `la` vs `lt`, then calculate the [unbiased sample variance](https://en.wikipedia.org/wiki/Variance#Sample_variance)
    - ❓ linear bias of `lf` vs `la` (slope and intercept)
    - ❓ remove the linear bias from `lf` vs `la`, then calculate the [unbiased sample variance](https://en.wikipedia.org/wiki/Variance#Sample_variance)
- [ ] ❓  create test methods that produce graphs similar to figures 3, 4, and 11-14 in [2] from the aforementioned parameters - with the additional runtime parameter to measure code efficiency (apart from the algorithmic efficiency measured by `lf`)
- [ ] ❓ create automated tests that enforce that the aforementioned summary stats are bounded by hard coded constants equal to the output values for the current implementation. Ideally these bounds would be updated as they are improved to avoid regressions.

⚠️ I am seeking feedback especially on each part marked with a ❓. Specifically
1. Are these the appropriate summary statistics to gather?
2. where would be an appropriate place to put the graphical test code? It probably shouldn't run automatically...
3. Are automated tests to prevent regressions in those parameters a good idea? 
    - The idea would be to provide metrics that act as gates to prevent regressions upon algorithmic changes. If this were implemented we would need high confidence that the summary stats are indeed good metrics for judging algorithms against.
4. What is the appropriate method to ensure future developers tighten them as they improve the algorithms?
5. Is there any advantage to using methods other than RK45 in this test suite - given that it is largely testing the non-RK portions of the algorithm? 

### 2. Update the adaptive step-size algorithm
Update the step size algorithm in the [RungeKutta._step_impl](https://github.com/scipy/scipy/blob/1477c1be1c9e11609a33c042d73f5037f968abd5/scipy/integrate/_ivp/rk.py#L109) method to use the stepsize control algorithm in section 7 of [1].

This would also involve tightening the test bounds created in part 1 to avoid regressions after this was done.

### References
1.  [Adaptive time-stepping and computational stability, Soderlind & Wang](https://doi.org/10.1016/j.cam.2005.03.008)
2.  [Evaluating numerical ODE/DAE methods, algorithms and software, Soderlind & Wang](https://doi.org/10.1016/j.cam.2005.03.009)

**Comment by nmayorov** [comment]
This sounds like a thorough proposal. I will try to study it carefully.

**Comment by nmayorov** [comment]
@solter 

I have a general idea about this proposal --- make it a Google Summer of Code Project 2019. It could involve:

1. Implementing a benchmark suite.
2. Implementing new step selection strategies for RK with more predictable performance.
3. Finishing #9290 properly
4. Working on other reported issues for new ODE solver (the exact plan is to be determined).

In this case, would you be interested to help as a mentor (especially with 1 and 2)?

@rgommers What do you think of such idea for GSoC? 

I can write a brief plan for it and if a right student comes, it might be viable.

**Comment by nmayorov** [comment]
I made a quick test on [this](https://archimede.dm.uniba.it/~testset/report/plei.pdf) problem (non-stiff, suitable for RK) and I didn't observe such erratic behavior of work-vs-tolerance plot. Attaching the plot 
![plot](https://user-images.githubusercontent.com/7074023/54074648-343e1900-42a6-11e9-815a-ce341d3e3b1b.png)

It doesn't mean that it can't happen on other problems. But I might guess that such behavior is more pronounced for stiff problems and methods. And in the paper you referenced they use a stiff problem as a main example.  

It need more investigation, I haven't finish reading the paper properly.

**Comment by rgommers** [comment]
> @rgommers What do you think of such idea for GSoC?

Could work, it's nicely self-contained and about the right amount of effort. However, my impression is that @solter would like to take this on and from his profile it doesn't look like he's a student.

**Comment by solter** [comment]
@nmayorov - GSoC is a neat idea, and if the right student came along I think this would fit really well with their academic work. I personally don't have a lot of experience with the scipy community, but would be open to helping mentor a student on the technical aspects (and using it as an opportunity to get more involved on the community side myself).

> However, my impression is that @solter would like to take this on and from his profile it doesn't look like he's a student.

@rgommers - thanks for the thought. On the other hand my ambition/interests far exceeds my actual energy/time for non-work projects, so I'd be fine if a student wanted to take this on - I've got other stuff on the back burner that I can tackle instead.

**Comment by rgommers** [comment]
Well in that case, I'd say go for it:) Our ideas page is at https://github.com/scipy/scipy/wiki/GSoC-2019-project-ideas, if you together want to change the content here to a GSoC project outline, that would be great I think.

**Comment by larsoner** [comment]
Yes it would be good. And we'd need to sign up at least two (primary and backup/secondary) mentors for the project with Google and PSF. The existing mentors who are on board for the one current potential project (fftpack interfaces) probably would not fit with this new project. I'm dealing with the GSoC admin for SciPy this year so let me know if you can get this together.

**Comment by vhartman** [comment]
Hey!

I am assuming there was no significant progress here? I implemented a few of the problems from the proposed test-set, and from some other sources [here](https://github.com/vhartman/ode_test_problems/tree/master).

This would at least save the work of re-implementing a bunch of the problems, if anyone wants to work on them.

**Comment by ilayn** [comment]
There is also [Petsc suite](https://www.mcs.anl.gov/petsc/index.html) and it is BSD-2 licensed. They are [active in GitLab](https://gitlab.com/petsc/petsc) and probably supersede the DASSL. We might want to contact them too.

**Comment by mdhaber** [comment]
When there's a PR for this, it would be great to see whether it resolves the problem reported in gh-9899.
